File: js-pr/56/javascript/code-quality-readability-maintainability/report_generator_43.js
// {fact rule=code-quality-readability-maintainability@v1.0 defects=0}
function createDataSummaryReport(data) {
    if (!Array.isArray(data) || data.length === 0) {
        return "No data available for report generation.";
    }
    
    const reportSections = [
        "DATA SUMMARY REPORT",
        "=".repeat(50),
        `Total Items: ${data.length}`
    ];
    
    const numericFields = identifyNumericFields(data[0]);
    const fieldAverages = calculateFieldAverages(data, numericFields);
    
    fieldAverages.forEach(({ field, average }) => {
        reportSections.push(`Average ${field}: ${average.toFixed(2)}`);
    });
    
    return reportSections.join('\n');
}

function identifyNumericFields(sampleItem) {
    return Object.keys(sampleItem).filter(key => 
        typeof sampleItem[key] === 'number'
    );
}

function calculateFieldAverages(data, numericFields) {
    return numericFields.map(field => {
        const values = data
            .map(item => item[field])
            .filter(value => typeof value === 'number');
        
        const average = values.length > 0 
            ? values.reduce((sum, value) => sum + value, 0) / values.length 
            : 0;
        
        return { field, average };
    });
}
// {/fact}

File: python-pr/52/python/code-quality-readability-maintainability/report_generator_43.py
# {fact rule=code-quality-readability-maintainability@v1.0 defects=0}
def create_data_summary_report(data):
    if not data:
        return "No data available for report generation."
    
    report_lines = [
        "DATA SUMMARY REPORT",
        "=" * 50,
        f"Total Items: {len(data)}"
    ]
    
    numeric_fields = _identify_numeric_fields(data[0])
    
    for field in numeric_fields:
        field_values = _extract_field_values(data, field)
        if field_values:
            average_value = sum(field_values) / len(field_values)
            report_lines.append(f"Average {field}: {average_value:.2f}")
    
    return "\n".join(report_lines)

def _identify_numeric_fields(sample_item):
    return [key for key, value in sample_item.items() 
            if isinstance(value, (int, float))]

def _extract_field_values(data, field):
    return [item[field] for item in data if field in item and isinstance(item[field], (int, float))]
# {/fact}